### PR TITLE
TFSlimClassifier, default deployments, and ETA lite installation

### PR DESCRIPTION
**Pull request by brimoor**
Change log:

**Major changes**
- adding an `eta.classifiers.TFSlimClassifier` interface that supports deploying models using the TF-Slim image classification library
- adding a number of popular pre-trained models for common TF-Slim architectures (inception, resnet, mobilenet, etc)
- adds the notion of a "default deployment config" to `eta.core.models.Model`. This allows us to document how one is supposed to deploy a model. There is also now an `eta.core.learning. load_default_deployment_model()` class that allows one to load a model of any name with a single line of code. This is useful for inspecting the behavior of a model
- adds an `eta.core.learning.HasDefaultDeploymentConfig` interface that allows models to declare that they have default deployments. `ModelConfig` instances can use this mixin to automatically load any settings from the default deployment whenever a model is loaded. This is useful, for example, to avoid having to re-specify the path to the labels map for a model every time you want to use it. The labels map will never change, so this cleans up our deployment code substantially!!
- adds a `eta.core.tfutils.UsesTFSession` mixin that classes can use to conveniently create `tf.Session` instances that inherit the global configuration settings from `eta.config.tf_config`. ALL CLASSES that use `tf.Session` should implement this mixin!!
- updating all existing uses of `tf.Session` to use the `UsesTFSession` mixin

**Minor changes**
- adding an `ETA lite` installation option that installs only what is necessary to use the `eta.core.package`, which notably excludes the gigantic `tensorflow/models` submodule
- adds support for writing patterns like `{{model}}` in a pipeline request that are then populated at build time via the `eta build --patterns model=my-model-name` syntax
- extends the list of supported image/video types to (hopefully!) be exhaustive

### CODE REVIEW

File: eta/detectors/tfmodels_detectors.py [hunk 1]
```diff
@@ -28,77 +28,81 @@
...
-    @staticmethod
-    def _build_graph(model_path):
-        tf_graph = tf.Graph()
-        with tf_graph.as_default():
-            graph_def = tf.GraphDef()
-            with tf.gfile.GFile(model_path, "rb") as f:
-                graph_def.ParseFromString(f.read())
-                tf.import_graph_def(graph_def, name="")
-        return tf_graph
+    def __enter__(self):
```
Comment (mattphotonman) [inline eta/detectors/tfmodels_detectors.py]: Why do we need these methods if it already inherits from `UsesTFSession`?

File: eta/classifiers/tfslim_classifiers.py [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+                network_name, None)
+            if self.output_name is None:
+                raise ValueError(
+                    "`output_name` was not provided and network `%s` was not "
+                    "found in default outputs map" % network_name)
+        self._output_op = self._graph.get_operation_by_name(
+            "prefix/" + self.output_name)
+
+        # Setup pre-processing
+        self._preprocessing_fcn = None
```
[the remark this answers is not in the document]
Comment (brimoor) [reply eta/classifiers/tfslim_classifiers.py]: style says to always declare all instance variables in the constructor of the class

File: eta/core/image.py [hunk 3]
```diff
@@ -45,7 +45,20 @@
 import eta.core.web as etaw
 
 
-SUPPORTED_IMAGE_FORMATS = [".png", ".jpg", ".jpeg", ".gif", ".tiff", ".bmp"]
+#
+# The file extensions of supported video files
```
[the remark this answers is not in the document]
Comment (brimoor) [reply eta/core/image.py]: oops! fixed

File: eta/core/models.py [hunk 4]
```diff
@@ -779,12 +797,16 @@ class Model(Serializable):
             location of the models_dir
         version: the version of the model (if any)
         description: a description of the model (if any)
+        default_deployment_config_dict: a dictionary representation of an
```
[the remark this answers is not in the document]
Comment (brimoor) [reply eta/core/models.py]: Aha! I did that initially, but creating a `ModelConfig` instance requires parsing a string of the form `eta.XXX.YYY`, which requires loading that module, and I didn't want to have to load all the modules for all kinds of crazy detectors/classifiers just to load the model manifest. Loading beefy modules can be slow.

So I decided to just load the dictionary here and only build the `ModelConfig` when a particular model is requested.

File: eta/core/tfutils.py [hunk 5]
```diff
@@ -36,19 +39,98 @@
...
+        super(TFModelCheckpoint, self).__init__(model_name)
+        self._sess = sess
+
+    def _load(self):
+        saver = tf.train.Saver()
+        saver.restore(self._sess, self.model_path)
+
+
+class UsesTFSession(object):
+    '''Base class for classes that use one or more `tf.Session`s.
```
[the remark this answers is not in the document]
Comment (brimoor) [reply eta/core/tfutils.py]: this is a mixin, good catch. fixed

File: eta/core/tfutils.py [hunk 6]
```diff
@@ -36,19 +39,98 @@
 TF_RECORD_EXTENSIONS = [".record", ".tfrecord"]
 
 
-def make_tf_session(config_proto=None):
+class TFModelCheckpoint(etam.PublishedModel):
```
[the remark this answers is not in the document]
Comment (brimoor) [reply eta/core/tfutils.py]: hmm, it's not really used as a mixin. There are no classes that subclass from this, it is just a small utility class that can load a model checkpoint into a `tf.session`

File: eta/detectors/tfmodels_detectors.py (hunk 1 above)
Comment (brimoor) [reply eta/detectors/tfmodels_detectors.py]: True. I just did it for clarity to emphasize that it should be used as a context manager

**Comment by brimoor** [comment]
@mattphotonman can you approve now? :)

**Comment by mattphotonman** [comment]
> @mattphotonman can you approve now? :)

I think you have to request a re-review first, I don't see a button to approve.

**Comment by brimoor** [comment]
@mattphotonman I re-requested your review. I think you should always be able to go to `Files changed -> Review changes -> Approve` without me doing that, though.
